core/Symbols.cc

```diff
@@ -812,6 +812,23 @@ SymbolRef Symbol::singletonClass(GlobalState &gs) {
     singletonInfo->setSuperClass(Symbols::todo());
     singletonInfo->setIsModule(false);
 
+    auto tp = gs.enterTypeMember(this->loc(), singleton, Names::Constants::AttachedClass(), Variance::CoVariant);
+
+    // Initialize the bounds of AttachedClass as top and bottom, as the upper
+    // bound will be updated to the result of `externalType` in the resolver's
+    // global pass.
+    tp.data(gs)->resultType = make_type<LambdaParam>(tp, Types::bottom(), Types::top());
+
+    // Only create the alias if it doesn't already exist. This happens when
+    // a singleton class has `singletonClass` called on it, as it already has a
+    // field named `AttachedClass`
+    auto alias = findMember(gs, Names::Constants::AttachedClass());
+    ENFORCE(!alias.exists() || isSingletonClass(gs));
+    if (!alias.exists()) {
+        alias = gs.enterStaticFieldSymbol(Loc::none(), selfRef, Names::Constants::AttachedClass());
+        alias.data(gs)->resultType = make_type<AliasType>(tp);
+    }
```

Comment:
Is this how all type_template's work? There's an alias on the attached class to the type member on the singleton class?

Reply:
https://github.com/sorbet/sorbet/blob/4bef9ab4572a5e4eef714cca9c7586de94eb051f/namer/namer.cc#L838-L839

huh til

Reply:
Yep! If you have a look at https://github.com/sorbet/sorbet/blob/trevor/attached-class/namer/namer.cc#L827-L840 you can see the alias being created for the usual `type_template` path.